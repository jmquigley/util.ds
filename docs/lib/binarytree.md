<a name="BinaryTree"></a>

## BinaryTree
Implements a binary tree structure using a Red/Black tree algorithm.

**Kind**: global class  

* [BinaryTree](#BinaryTree)
    * [.breadth](#BinaryTree+breadth) ⇒ <code>Array.&lt;T&gt;</code>
    * [.height](#BinaryTree+height) ⇒ <code>number</code>
    * [.inorder](#BinaryTree+inorder) ⇒ <code>Array.&lt;T&gt;</code>
    * [.postorder](#BinaryTree+postorder) ⇒ <code>Array.&lt;T&gt;</code>
    * [.preorder](#BinaryTree+preorder) ⇒ <code>Array.&lt;T&gt;</code>
    * [.breadthSearch(data, node)](#BinaryTree+breadthSearch) ⇒ <code>boolean</code>
    * [.clear()](#BinaryTree+clear)
    * [.contains(data)](#BinaryTree+contains) ⇒ <code>boolean</code>
    * [.find(key)](#BinaryTree+find) ⇒ <code>T</code>
    * [.insert(data)](#BinaryTree+insert)
    * [.remove(data)](#BinaryTree+remove)
    * [.removeFirst()](#BinaryTree+removeFirst) ⇒ <code>T</code>
    * [.removeLast()](#BinaryTree+removeLast) ⇒ <code>T</code>
    * [._findNode(data)](#BinaryTree+_findNode) ⇒ <code>Node.&lt;T&gt;</code>
    * [._leftRotate()](#BinaryTree+_leftRotate)
    * [._maximum(node)](#BinaryTree+_maximum) ⇒ <code>Node.&lt;T&gt;</code>
    * [._minimum(node)](#BinaryTree+_minimum) ⇒ <code>Node.&lt;T&gt;</code>
    * [._rightRotate()](#BinaryTree+_rightRotate)
    * [._successor(node)](#BinaryTree+_successor) ⇒ <code>Node.&lt;T&gt;</code>

<a name="BinaryTree+breadth"></a>

### binaryTree.breadth ⇒ <code>Array.&lt;T&gt;</code>
Performs a breadth first traversal of the tree and saves it to an array
of T type (of the tree).  The array of data is returned.  Don't use this
with a large tree.

**Kind**: instance property of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>Array.&lt;T&gt;</code> - an array of all elements in the tree in breadth order  
<a name="BinaryTree+height"></a>

### binaryTree.height ⇒ <code>number</code>
**Kind**: instance property of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>number</code> - computes and returns the height of the tree.  
<a name="BinaryTree+inorder"></a>

### binaryTree.inorder ⇒ <code>Array.&lt;T&gt;</code>
**Kind**: instance property of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>Array.&lt;T&gt;</code> - the results of an inorder traversal of the tree.  The
results are stored in an array and returned.  
<a name="BinaryTree+postorder"></a>

### binaryTree.postorder ⇒ <code>Array.&lt;T&gt;</code>
**Kind**: instance property of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>Array.&lt;T&gt;</code> - the results of a postorder traversal of the tree.  The
results are stored in an array and returned.  
<a name="BinaryTree+preorder"></a>

### binaryTree.preorder ⇒ <code>Array.&lt;T&gt;</code>
**Kind**: instance property of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>Array.&lt;T&gt;</code> - the results of a preorder traversal of the tree.  The
results are stored in an array and returned.  
<a name="BinaryTree+breadthSearch"></a>

### binaryTree.breadthSearch(data, node) ⇒ <code>boolean</code>
Peforms a breadth first search against the tree.  Generally this is not
the best way to search the tree (use contains to determine if a key is
in it).

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>boolean</code> - true if the item is found in the tree, otherwise
false is returned.  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>T</code> | the data element to search for (based on the data type of the tree) |
| node | <code>T</code> | the starting node for the search.  This is root by default |

<a name="BinaryTree+clear"></a>

### binaryTree.clear()
Initializes the object to an emtpy state.  This can be used to
quickly empty the tree and start over.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
<a name="BinaryTree+contains"></a>

### binaryTree.contains(data) ⇒ <code>boolean</code>
Performs a typically binary search through the tree.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>boolean</code> - true if the item is found in the tree, otherwise
false is returned.  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>T</code> | the data element to search for (based on the data type of the tree) |

<a name="BinaryTree+find"></a>

### binaryTree.find(key) ⇒ <code>T</code>
Searches the tree for an element.  If it is found, then the data element
associated with that node is returned (not the node).  When used with a
primative type this is not useful as the key and the value found would
be the same.  This is helpful when T is a complex object with a custom
Comparator.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>T</code> - the full data element within this tree.  If it is not found,
then null is returned.  

| Param | Type | Description |
| --- | --- | --- |
| key | <code>T</code> | a search key to look for in the tree. |

<a name="BinaryTree+insert"></a>

### binaryTree.insert(data)
Inserts a data element into the tree.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>T</code> | the data element to insert into the tree |

<a name="BinaryTree+remove"></a>

### binaryTree.remove(data)
Removes the given data value from the tree.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>T</code> | the data value to remove |

<a name="BinaryTree+removeFirst"></a>

### binaryTree.removeFirst() ⇒ <code>T</code>
Special case function to quickly find and remove the first item in the
tree.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>T</code> - the data element that was first and removed from the tree.  
<a name="BinaryTree+removeLast"></a>

### binaryTree.removeLast() ⇒ <code>T</code>
Special case function to quickly find and remove the last item in the
tree.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>T</code> - the data element that was last and removed from the tree.  
<a name="BinaryTree+_findNode"></a>

### binaryTree._findNode(data) ⇒ <code>Node.&lt;T&gt;</code>
Searches the tree for a specific node within the tree.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>Node.&lt;T&gt;</code> - if the data is found, then the node that holds it is
returned.  If it is not found, then nil is returned.  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>T</code> | the data value to search for in the tree. |

<a name="BinaryTree+_leftRotate"></a>

### binaryTree._leftRotate()
Localized left rotation of nodes.  This is a public function but is private
by convention (for testing).  Generally not called as part of the api.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
<a name="BinaryTree+_maximum"></a>

### binaryTree._maximum(node) ⇒ <code>Node.&lt;T&gt;</code>
Searches a tree from a given node for the maximum value in that
(sub)tree.  Note that the property `.largest` can be used to
quickly retrieve the largest value in the tree.  This is really used
to recompute the maximum value when it is removed from the tree.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>Node.&lt;T&gt;</code> - the largest node in the (sub)tree.  

| Param | Type | Description |
| --- | --- | --- |
| node | <code>Node.&lt;T&gt;</code> | the node location to start the search.  By default this is the root node if no node is given. |

<a name="BinaryTree+_minimum"></a>

### binaryTree._minimum(node) ⇒ <code>Node.&lt;T&gt;</code>
From a node, searches a tree or subtree for the minimum value in that
(sub)tree.  Note that the property `.smallest` can be used to
quickly retrieve the smallest value in the tree.  This is really used
to recompute the minimum value when it is removed from the tree.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>Node.&lt;T&gt;</code> - the smallest node in the (sub)tree.  

| Param | Type | Description |
| --- | --- | --- |
| node | <code>Node.&lt;T&gt;</code> | the node location to start the search.  By default this is the root node if no node is given. |

<a name="BinaryTree+_rightRotate"></a>

### binaryTree._rightRotate()
Localized right rotation of nodes.  This is a public function but is private
by convention (for testing).  Generally not called as part of the api.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
<a name="BinaryTree+_successor"></a>

### binaryTree._successor(node) ⇒ <code>Node.&lt;T&gt;</code>
The successor of a node is the node with the smallest key greater than
node.data.

**Kind**: instance method of [<code>BinaryTree</code>](#BinaryTree)  
**Returns**: <code>Node.&lt;T&gt;</code> - a reference to the successor node.  

| Param | Type | Description |
| --- | --- | --- |
| node | <code>Node.&lt;T&gt;</code> | the node location to start the search for a successor. |

