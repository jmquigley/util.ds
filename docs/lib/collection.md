<a name="Collection"></a>

## Collection
**Kind**: global class  

* [Collection](#Collection)
    * [new Collection(comparator)](#new_Collection_new)
    * [.back](#Collection+back) ⇒ <code>T</code>
    * [.empty](#Collection+empty) ⇒ <code>boolean</code>
    * [.first](#Collection+first) ⇒ <code>T</code>
    * [.front](#Collection+front) ⇒ <code>T</code>
    * [.last](#Collection+last) ⇒ <code>T</code>
    * [.length](#Collection+length) ⇒ <code>number</code>
    * [.nil](#Collection+nil) ⇒ <code>Node.&lt;T&gt;</code>
    * [.root](#Collection+root) ⇒ <code>Node.&lt;T&gt;</code>
    * [.size](#Collection+size) ⇒ <code>number</code>
    * [.clear()](#Collection+clear)
    * [.contains(obj)](#Collection+contains) ⇒ <code>boolean</code>
    * [.isEmpty()](#Collection+isEmpty) ⇒ <code>boolean</code>

<a name="new_Collection_new"></a>

### new Collection(comparator)
Base class constructor for all collection classes.

**Params**

- comparator <code>function</code> <code> = </code> - a comparator function used for searching within
the container.

<a name="Collection+back"></a>

### collection.back ⇒ <code>T</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>T</code> - the last (max) data element from the tree.  
<a name="Collection+empty"></a>

### collection.empty ⇒ <code>boolean</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>boolean</code> - return true if the list has no nodes, otherwise false  
<a name="Collection+first"></a>

### collection.first ⇒ <code>T</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>T</code> - the first (min) data element from the tree.  
<a name="Collection+front"></a>

### collection.front ⇒ <code>T</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>T</code> - the front (min) data element from the tree.  
<a name="Collection+last"></a>

### collection.last ⇒ <code>T</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>T</code> - the last (max) data element from the tree.  
<a name="Collection+length"></a>

### collection.length ⇒ <code>number</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>number</code> - the number of nodes in this collection  
<a name="Collection+nil"></a>

### collection.nil ⇒ <code>Node.&lt;T&gt;</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>Node.&lt;T&gt;</code> - the reference to the nil sentinel  
<a name="Collection+root"></a>

### collection.root ⇒ <code>Node.&lt;T&gt;</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>Node.&lt;T&gt;</code> - the front/first node in the collection  
<a name="Collection+size"></a>

### collection.size ⇒ <code>number</code>
**Kind**: instance property of [<code>Collection</code>](#Collection)  
**Returns**: <code>number</code> - the number of nodes in this collection  
<a name="Collection+clear"></a>

### collection.clear()
Initializes the object to an empty state

**Kind**: instance method of [<code>Collection</code>](#Collection)  
<a name="Collection+contains"></a>

### collection.contains(obj) ⇒ <code>boolean</code>
Checks the container for the existence of a given object.  This is
a simple linear search.

**Kind**: instance method of [<code>Collection</code>](#Collection)  
**Returns**: <code>boolean</code> - true if the item is found, otherwise false.  
**Params**

- obj <code>Object</code> - the item to find in the container.

<a name="Collection+isEmpty"></a>

### collection.isEmpty() ⇒ <code>boolean</code>
Checks if the container is empty.

**Kind**: instance method of [<code>Collection</code>](#Collection)  
**Returns**: <code>boolean</code> - true if the container is empty, otherwise false.  
