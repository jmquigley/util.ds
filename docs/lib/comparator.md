<a name="defaultComparatorFn"></a>

## defaultComparatorFn(o1, o2) â‡’ <code>number</code>
This comparator is used to handle default comparasons for built in types

It returns:

0  - if the objects are equal
1  - if o1 > o2
-1 - if o1 < o2

**Kind**: global function  
**Returns**: <code>number</code> - the result of the comparison (above)  

| Param | Type | Description |
| --- | --- | --- |
| o1 | <code>T</code> | the left hand argument to compare |
| o2 | <code>T</code> | the right hand argument to compare |

