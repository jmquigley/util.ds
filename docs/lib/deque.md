<a name="Deque"></a>

## Deque
A double ended Queue class

**Kind**: global class  

* [Deque](#Deque)
    * [new Deque(maxSize, arr, comparator)](#new_Deque_new)
    * [.overflow](#Deque+overflow) ⇒ <code>boolean</code>
    * [.enqueue(data)](#Deque+enqueue)
    * [.pushFront(data)](#Deque+pushFront)
    * [.pushBack(data)](#Deque+pushBack)
    * [.popFront()](#Deque+popFront) ⇒ <code>Object</code>
    * [.popBack()](#Deque+popBack) ⇒ <code>Object</code>

<a name="new_Deque_new"></a>

### new Deque(maxSize, arr, comparator)
The deque can be unlimited in size (default) or set to a maxium size
when constructed.  When the max size is exceeded, then the front item is
automatically removed from the queue and the new item is placed in the
queue (depending on  which type of insert is calld).


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| maxSize | <code>number</code> | <code>0</code> | The maximum size for this queue. |
| arr | <code>Array.&lt;T&gt;</code> | <code>[</code> | An array of initial input values |
| comparator | <code>function</code> | <code></code> | a comparator function used for searching within the container. |

<a name="Deque+overflow"></a>

### deque.overflow ⇒ <code>boolean</code>
Checks the current internals for an overflow condition.  This occurs
when the maxSize storage size will be exceeded on the next insert
operation.

**Kind**: instance property of [<code>Deque</code>](#Deque)  
**Returns**: <code>boolean</code> - true if in an overflow condition, otherwise false.  
<a name="Deque+enqueue"></a>

### deque.enqueue(data)
Adds an item to the end of the queue.  Checks for an overflow condition
and acts if one is detected.

**Kind**: instance method of [<code>Deque</code>](#Deque)  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>Object</code> | the data item to add to the queue. |

<a name="Deque+pushFront"></a>

### deque.pushFront(data)
Inserts a data element to the front of the queue.  Checks for an
overflow condition and acts if one is detected.

**Kind**: instance method of [<code>Deque</code>](#Deque)  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>Object</code> | the data item to add to the queue. |

<a name="Deque+pushBack"></a>

### deque.pushBack(data)
Adds an item to the end of the queue.  Checks for an overflow condition
and acts if one is detected.

**Kind**: instance method of [<code>Deque</code>](#Deque)  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>Object</code> | the data item to add to the queue. |

<a name="Deque+popFront"></a>

### deque.popFront() ⇒ <code>Object</code>
Returns the first item from the queue and removes it.

**Kind**: instance method of [<code>Deque</code>](#Deque)  
**Returns**: <code>Object</code> - the data value on the front of the queue.  
<a name="Deque+popBack"></a>

### deque.popBack() ⇒ <code>Object</code>
Retrieves the last item from the queue and removes it.

**Kind**: instance method of [<code>Deque</code>](#Deque)  
**Returns**: <code>Object</code> - the data value on the back of the queue.  
