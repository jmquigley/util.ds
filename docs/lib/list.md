<a name="List"></a>

## List
A doubly linked list structure.

**Kind**: global class  

* [List](#List)
    * [.array](#List+array) ⇒ <code>Array.&lt;T&gt;</code>
    * [.reverse](#List+reverse) ⇒ <code>Array.&lt;T&gt;</code>
    * [.find(key)](#List+find) ⇒ <code>T</code>
    * [.insert(data, idx)](#List+insert)
    * [.remove(data, [idx])](#List+remove) ⇒ <code>T</code>
    * [._getNodeByValue(data)](#List+_getNodeByValue) ⇒ <code>Node.&lt;T&gt;</code>
    * [._getNodeByIndex(idx)](#List+_getNodeByIndex) ⇒ <code>Node.&lt;T&gt;</code>

<a name="List+array"></a>

### list.array ⇒ <code>Array.&lt;T&gt;</code>
**Kind**: instance property of [<code>List</code>](#List)  
**Returns**: <code>Array.&lt;T&gt;</code> - an inorder array of all elements in the list  
<a name="List+reverse"></a>

### list.reverse ⇒ <code>Array.&lt;T&gt;</code>
**Kind**: instance property of [<code>List</code>](#List)  
**Returns**: <code>Array.&lt;T&gt;</code> - a reversed array of all elements in the list  
<a name="List+find"></a>

### list.find(key) ⇒ <code>T</code>
Searches the list for an element.  If it is found, then the data element
associated with that node is returned (not the node).  When used with a
primative type this is not useful as the key and the value found would
be the same.  This is helpful when T is a complex object with a custom
Comparator.

**Kind**: instance method of [<code>List</code>](#List)  
**Returns**: <code>T</code> - the full data element within this list.  If it is not found,
then null is returned.  

| Param | Type | Description |
| --- | --- | --- |
| key | <code>T</code> | a search key to look for in the list. |

<a name="List+insert"></a>

### list.insert(data, idx)
Inserts a data value into a linked list.  The default operation is to
insert into the end of the list.

**Kind**: instance method of [<code>List</code>](#List)  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>T</code> | the data item to insert into the list |
| idx | <code>number</code> | the index position where the item will be inserted. this number be List.FRONT, List.BACK, or any other number. |

<a name="List+remove"></a>

### list.remove(data, [idx]) ⇒ <code>T</code>
Removes an element from the list.

**Kind**: instance method of [<code>List</code>](#List)  
**Returns**: <code>T</code> - the data element that was deleted  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| data | <code>T</code> |  | the data item removed from the list |
| [idx] | <code>number</code> | <code></code> | the position index to remove from the list |

<a name="List+_getNodeByValue"></a>

### list._getNodeByValue(data) ⇒ <code>Node.&lt;T&gt;</code>
Iterates through the list and find the node associated with a key
value T.

**Kind**: instance method of [<code>List</code>](#List)  
**Returns**: <code>Node.&lt;T&gt;</code> - a reference to the node that was found.  If a node
is not found, then null is returned.  

| Param | Type | Description |
| --- | --- | --- |
| data | <code>T</code> | the key value or index to search for in the list. |

<a name="List+_getNodeByIndex"></a>

### list._getNodeByIndex(idx) ⇒ <code>Node.&lt;T&gt;</code>
Moves to a position within the list and finds the node associated
with that index.

**Kind**: instance method of [<code>List</code>](#List)  
**Returns**: <code>Node.&lt;T&gt;</code> - a reference to a node a the given position.  If the
node is not found, then null is returned.  

| Param | Type | Description |
| --- | --- | --- |
| idx | <code>number</code> | the location within the list to find |

