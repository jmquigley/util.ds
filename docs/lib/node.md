<a name="Node"></a>

## Node
The data nodes used within a collection.  This generally would not be
used outside of these collection classes.

**Kind**: global class  
