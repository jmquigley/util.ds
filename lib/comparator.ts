/**
 * A comparator function signature.  This is used to perform a comparison
 * operation between two objects.
 */
export type Comparator<T> = (o1: T, o2: T) => number;

/**
 * This comparator is used to handle default comparasons for built in types
 *
 * It returns:
 *
 * 0  - if the objects are equal
 * 1  - if o1 > o2
 * -1 - if o1 < o2
 *
 * @param o1 {T} the left hand argument to compare
 * @param o2 {T} the right hand argument to compare
 * @return {number} the result of the comparison (above)
 */
export function defaultComparatorFn<T>(o1: T, o2: T): number {
	if (o1 === o2) {
		return 0;
	} else if (o1 > o2) {
		return 1;
	}

	return -1;
}
