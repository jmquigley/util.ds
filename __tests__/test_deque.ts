'use strict';

import {Deque} from '../index';
import {Comparator} from '../lib/comparator';

interface TestData {
	item1: string;
	item2: number;
}

test('Create an emtpy Deque', () => {
	const dq = new Deque<number>();

	expect(dq).toBeDefined();
	expect(dq.isEmpty()).toBe(true);
	expect(dq.empty).toBe(true);
	expect(dq.front === null).toBe(true);
	expect(dq.back === null).toBe(true);
	expect(dq.popFront() == null).toBe(true);
	expect(dq.popBack() == null).toBe(true);
	expect(dq).toMatchSnapshot();
});

test('Add/Remove items from the front of the deque', () => {
	const dq = new Deque<number>();
	const n: number = 5;

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	for (let i: number = 0; i < n; i++) {
		dq.pushFront(i);
	}

	expect(dq.size).toBe(5);

	for (let i: number = n; i > 0; i--) {
		expect(dq.length).toBe(i);
		expect(dq.front).toBe(i - 1);
		expect(dq.popFront()).toBe(i - 1);
	}

	expect(dq.empty).toBe(true);
});

test('Add/Remove items from the back of the deque', () => {
	const dq = new Deque<number>();
	const n: number = 5;

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	for (let i: number = 0; i < n; i++) {
		dq.pushBack(i);
	}

	expect(dq.size).toBe(5);

	for (let i: number = n; i > 0; i--) {
		expect(dq.length).toBe(i);
		expect(dq.back).toBe(i - 1);
		expect(dq.popBack()).toBe(i - 1);
	}

	expect(dq.empty).toBe(true);
});

test('Test deque insert event (pushFront)', (done) => {
	const dq = new Deque<number>();

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	const n: number = 100;
	dq.on('insert', (data: any) => {
		expect(data).toBe(n);
		done();
	});

	dq.pushFront(n);
});

test('Test deque remove event (popFront)', (done) => {
	const dq = new Deque<number>();

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	const n: number = 100;
	dq.on('remove', (data: any) => {
		expect(data).toBe(n);
		done();
	});

	dq.pushFront(n);
	dq.popFront();
});

test('Test deque insert event (pushBack)', (done) => {
	const dq = new Deque<number>();

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	const n: number = 100;
	dq.on('insert', (data: any) => {
		expect(data).toBe(n);
		done();
	});

	dq.pushBack(n);
});

test('Test deque remove event (popBack)', (done) => {
	const dq = new Deque<number>();

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	const n: number = 100;
	dq.on('remove', (data: any) => {
		expect(data).toBe(n);
		done();
	});

	dq.pushBack(n);
	dq.popBack();
});

test('Test size limited Deque (front)', () => {
	const n: number = 5;
	const dq = new Deque<number>(n);

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	for (let i: number = 0; i < n; i++) {
		dq.enqueue(i);
	}

	expect(dq.size).toBe(n);
	dq.pushFront(n);
	expect(dq.size).toBe(n);

	const arr = dq.drain();
	expect(arr[0]).toBe(5);
	expect(arr[1]).toBe(1);
	expect(arr[2]).toBe(2);
	expect(arr[3]).toBe(3);
	expect(arr[4]).toBe(4);

	expect(dq.empty).toBe(true);
});

test('Test size limited Deque (back)', () => {
	const n: number = 5;
	const dq = new Deque<number>(n);

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	for (let i: number = 0; i < n; i++) {
		dq.enqueue(i);
	}

	expect(dq.size).toBe(n);
	dq.enqueue(n);
	expect(dq.size).toBe(n);

	const arr = dq.drain();
	expect(arr[0]).toBe(1);
	expect(arr[1]).toBe(2);
	expect(arr[2]).toBe(3);
	expect(arr[3]).toBe(4);
	expect(arr[4]).toBe(5);

	expect(dq.empty).toBe(true);
});

test('Test size limited Deque (pushFront) event', (done) => {
	const n: number = 5;
	const dq = new Deque<number>(n);

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	for (let i: number = 0; i < n; i++) {
		dq.enqueue(i);
	}

	expect(dq.size).toBe(n);

	dq.on('remove', (data: any) => {
		expect(dq.size).toBe(n - 1);
		expect(data).toBe(0);
		done();
	});

	dq.pushFront(n + 1);
});

test('Test size limited Deque (pushBack) event', (done) => {
	const n: number = 5;
	const dq = new Deque<number>(n);

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	for (let i: number = 0; i < n; i++) {
		dq.enqueue(i);
	}

	expect(dq.size).toBe(n);

	dq.on('remove', (data: any) => {
		expect(dq.size).toBe(n - 1);
		expect(data).toBe(0);
		done();
	});

	dq.pushBack(n + 1);
});

test('Test the contains function on empty deque', () => {
	const dq = new Deque<number>();

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);
	expect(!dq.contains(999)).toBe(true);
});

test('Test the contains function for a deque', () => {
	const dq = new Deque<number>();

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	const n: number = 100;
	for (let i = 0; i < n; i++) {
		dq.enqueue(i);
	}

	expect(dq.contains(1)).toBe(true);
	expect(dq.contains(10)).toBe(true);
	expect(!dq.contains(999)).toBe(true);
});

test('Test of the deque contains function with custom comparator', () => {
	const fn: Comparator<TestData> = (o1: TestData, o2: TestData): number => {
		if (o1.item1 === o2.item1 && o1.item2 === o2.item2) {
			return 0;
		} else if (o1.item1 > o2.item1 && o1.item2 > o2.item2) {
			return 1;
		}

		return -1;
	};

	const dq = new Deque<TestData>(10, null, fn);

	expect(dq).toBeDefined();
	expect(dq.empty).toBe(true);

	dq.enqueue({item1: 'abc', item2: 0});
	dq.enqueue({item1: 'abc', item2: 1});
	dq.enqueue({item1: 'def', item2: 0});
	dq.enqueue({item1: 'ghi', item2: 0});
	dq.enqueue({item1: 'jkl', item2: 0});
	dq.enqueue({item1: 'mno', item2: 0});

	expect(dq.contains({item1: 'abc', item2: 1})).toBe(true);
	expect(dq.contains({item1: 'ghi', item2: 0})).toBe(true);
	expect(dq.contains({item1: 'mno', item2: 0})).toBe(true);
	expect(!dq.contains({item1: 'cat', item2: 0})).toBe(true);
	expect(!dq.contains({item1: 'dog', item2: 0})).toBe(true);
});
