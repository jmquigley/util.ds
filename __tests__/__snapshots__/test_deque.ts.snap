// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create an emtpy Deque 1`] = `
Deque {
  "_comparator": [Function],
  "_end": null,
  "_events": Object {},
  "_eventsCount": 0,
  "_first": null,
  "_last": null,
  "_length": 0,
  "_maxListeners": undefined,
  "_maxSize": 0,
  "_nil": Node {
    "_options": Object {
      "children": Array [],
      "color": 1,
      "data": null,
      "id": null,
      "left": null,
      "parent": null,
      "parentId": null,
      "right": null,
    },
    "children": Array [],
    "color": 1,
    "data": null,
    "id": null,
    "left": null,
    "parent": null,
    "parentId": null,
    "right": null,
  },
  "_root": null,
  "push": [Function],
}
`;
